fusion/FSL.md

```diff
@@ -0,0 +1,1220 @@
+C.  The left column is the encoding, the comment is the Mavis unique identifier 
+for each instruction which is independent of the operands.
+
+```
+0x0512     c.slli x10,4    ; 0xf
+0x8111     c.srli x10,4    ; 0x13
+```
+
+If you do not require constraints on the operands and therefore have no
+need to refer to the operands by name, you can simply specify the UIDs 
+as a list, one per line.
+
+```
+ioput iop1
+
+sequence seq1(iop1,rv64g) {
+  0xf
+  0x13
+}
+```
+
+With the UID sequence expression there are no operand specific 
+constraints and therefore the conversion clause must handle any legal
+combinations of the two rd and two constants.
+
+This results in a transform to a generalized fusion op of a shift left 
+followed by right and not the zero extension expressed by x10,4.
+
+If the constraints clause implements operand contraints the instruction
+sequence should be expressed using the abstract assembly syntax so the
+operands can be referenced by name.
+
+The expression below takes advantage of positional constraints:
+
+```
+ioput iop1
+
+sequence seq1(iop1,rv64g) {
+  c.slli g1,c1
+  c.srli g1,c1
+}
+```
+
+In this case the operands labeled g1 are, by association, required to be
+the same value, similarly with the constant operands.
+
+Another option would be to fully enumerate the operands and let the
+constraints clause enforce the required limits on operands.
+
+```
+ioput iop1
+
+sequence seq1(iop1,rv64g) {
+  c.slli g1,c1
+  c.srli g2,c2
+}
+```
+
+Note both the later two cases can be made equivalent by construction of
+the equivalent constraints in the clause.
+
+Now in both cases the transformation expresses a fusion operation for
+zero extention of the common rd register by the common constant.
+
+```
+g1 = g2
+c1 = c2
+```
+
+In some cases positional constraints can save development time at no 
+expense to performance.
+
+The example sequences above have an implied strict ordering. There are 
+sequence pragmas to relax the strictness of sequence matching. 
+
+The required pragma, \_req\_, indicates that an unspecified instruction
+is required in that position. The \_req\_ case does not constrain 
+what instruction can be present in the gap. A trailing integer 
+can specify more than one required instruction.
+
+The optional pragma, \_opt\_, indicates that the match will not be 
+rejected if there is an instruction or not in this position. A
+trailing integer can be used to specify up to N optional instructions.
+
+```
+_opt_ 2    indicates 0 - 2 instructions will match
+_req_ 2    indicates there must be 2 instructions in this position
+           in the sequence. 
+```
+
+The trailing integer is optional in the syntax:
+```
+_opt_ and _opt_ 1 are equivalent
+_req_ and _req_ 1 are equivalent
+```
+
+```
+ioput iop1
+
+sequence seq1(iop1,rv64g) {
+  c.lui    g1, c1
+  c.addi   g1, g1, c2
+  _req_ 2
+  c.xor    g2, g2, g1
+  c.slli   g2, g2, c3
+  _opt_ 2
+  c.srli   g2,     c3
+}
+```
+
```

Comment:
Is there a way to fuse one or more classes of instruction with other instructions? For example, is there a way to define a `logical immediate` class of instructions for `andi`, `ori`, `xori` and define a sequence in terms of that class, instead of individual instructions?

Reply:
There is no way to do this in the current grammar. Let me give it some thought.

Reply:
I think the proper way to handle the need is to provide syntax for composition of sequences from other sequences rather than rely on meta-data based classifications. 

I'll do this in a future version.

Reply:
Can we please capture this feature/need in an issue? Or if you would like, may be create one issue to capture various future improvements of FSL?